node/src/exchange.rs

```diff
@@ -0,0 +1,98 @@
+use std::io;
+
+use async_trait::async_trait;
+use celestia_proto::p2p::pb::{HeaderRequest, HeaderResponse};
+use futures::prelude::*;
+use futures::{AsyncRead, AsyncWrite};
+use libp2p::request_response::{self, Codec, ProtocolSupport};
+use libp2p::StreamProtocol;
+use prost::Message;
+
+use crate::utils::stream_protocol_id;
+
+/// Max request size in bytes
+const REQUEST_SIZE_MAXIMUM: u64 = 1024;
+/// Max response size in bytes
+const RESPONSE_SIZE_MAXIMUM: u64 = 10 * 1024 * 1024;
+
+pub type Behaviour = request_response::Behaviour<HeaderCodec>;
+pub type Event = request_response::Event<HeaderRequest, HeaderResponse>;
+
+/// Create a new [`Behaviour`]
+pub fn new_behaviour(network: &str) -> Behaviour {
+    Behaviour::new(
+        [(
+            stream_protocol_id(network, "/header-ex/v0.0.3"),
+            ProtocolSupport::Full,
+        )],
+        request_response::Config::default(),
+    )
+}
+
+#[derive(Clone, Copy, Debug, Default)]
+pub struct HeaderCodec;
+
+#[async_trait]
+impl Codec for HeaderCodec {
+    type Protocol = StreamProtocol;
+    type Request = HeaderRequest;
+    type Response = HeaderResponse;
+
+    async fn read_request<T>(&mut self, _: &Self::Protocol, io: &mut T) -> io::Result<Self::Request>
+    where
+        T: AsyncRead + Unpin + Send,
+    {
+        let mut vec = Vec::new();
+
+        io.take(REQUEST_SIZE_MAXIMUM).read_to_end(&mut vec).await?;
```

Comment:
We could try to utilize `prost::decode_length_delimiter` and `prost::length_delimiter_len` to just pre-alloc the needed vector that we'll read to.
Commenting just to open the issue